### PR TITLE
[distribution] Fix expanded mvn and lowrankmvn

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR fixes an issue of the slowness expanded MVN.

A notebook to show the problem is [here](https://gist.github.com/fehiepsi/b15ac2978f1045d6d96b1d35b640d742). Basically, mvn's sample and log_prob have expensive computations based on `cholesky` and `trtrs`. We can save a lot of computation based on caching the unbroadcasted version of `scale_tril` (or `cov_diag`, `cov_factor` in lowrank mvn).
When expanding, this cached tensor should not be expanded together with other arguments.

Ref: https://github.com/uber/pyro/issues/1586

cc @neerajprad @fritzo

### CODE REVIEW

**Review by fritzo — COMMENTED** [review]
This looks fine, but I wonder if it would be simpler to implement an `unexpand` method
```py
def unexpand(x):
    return x[tuple(slice(None) if stride else slice(None, 1)
                   for stride in x.stride())]
```
and use this before each operation
```py
unbroadcasted_cov = unexpand(self.cov)
...
```

File: torch/distributions/kl.py [hunk 1]
```diff
@@ -300,19 +300,24 @@ def _kl_lowrankmultivariatenormal_lowrankmultivariatenormal(p, q):
         raise ValueError("KL-divergence between two Low Rank Multivariate Normals with\
                           different event shapes cannot be computed")
 
-    term1 = (_batch_lowrank_logdet(q.cov_factor, q.cov_diag, q._capacitance_tril) -
-             _batch_lowrank_logdet(p.cov_factor, p.cov_diag, p._capacitance_tril))
-    term3 = _batch_lowrank_mahalanobis(q.cov_factor, q.cov_diag, q.loc - p.loc,
+    term1 = (_batch_lowrank_logdet(q._unbroadcasted_cov_factor, q._unbroadcasted_cov_diag,
```
Comment (neerajprad) [inline torch/distributions/kl.py]: Just to confirm - will this give the correctly shaped kl (i.e. with the expanded shape)?

**Comment by fehiepsi** [comment]
@fritzo If the `unexpand` operator is cheap, then I agree that there is no need to store `unbroadcasted_scaled_tril`, `unbroadcasted_cov_factor`, `unbroadcasted_cov_diag`. Both are fine to me. If we use `unexpand`, then we might have to expand again to match the event_shape though (I am not so sure about this, as I imagine, `unexpand` will "unexpand" a tensor to its root tensor, which might not fit the distribution's event shape).

File: torch/distributions/kl.py (hunk 1 above)
Comment (fehiepsi) [reply torch/distributions/kl.py]: @neerajprad The `term3` involves `q.loc` and `p.loc`, which always match the distribution shapes (batch + event).

**Comment by neerajprad** [comment]
> This looks fine, but I wonder if it would be simpler to implement an unexpand method

One advantage to defining `unexpand` is that we could use it in other distributions `log_prob` methods too, if needed. 

> Both are fine to me. If we use unexpand, then we might have to expand again to match the event_shape though

I think it would be fine in most cases, but it is safe only if we call `.contiguous` on all the parameters in the distribution's `__init__`.

**Comment by fritzo** [comment]
I don't see why `unexpand` and `expand` would require `.contiguous()`.

Why don't we try to get this PR merged, then see if we can get a jittable C++ implementation of `unexpand` after 1.0 release and refactor some distribution methods to use it.

**Review by neerajprad — APPROVED** [review]

**Comment by fehiepsi** [comment]
@apaszke Is there any requirement to have this for 1.0? I'll try my best to full fill it. :)

**Review by apaszke — APPROVED** [review]
Haven't reviewed the math, but happy to stamp if @fritzo and @neerajprad accept.

@soumith do we need to do anything special to include this in 1.0?

**Comment by soumith** [comment]
i'll make sure it gets into 1.0

**Pull request closed without merging**